walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,60 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.Iterator;
+
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    List<TimeRange> availableTimes = new ArrayList<>();
+    long requestDuration = request.getDuration();
+
+    // edge cases
+    if (requestDuration > 1440) {
+        return availableTimes;
+    }
+
+    TimeRange timeLeft = TimeRange.fromStartDuration(0, 24 * 60);
+
+    for (Event e : events) {
+        TimeRange currEvent = e.getWhen();
+
+        if (timeLeft.overlaps(currEvent) && checkDuplicateAttendees(request.getAttendees(), e.getAttendees())) {
+            TimeRange newSlot = TimeRange.fromStartEnd(timeLeft.start(), currEvent.start(), false);
+            timeLeft = TimeRange.fromStartEnd(currEvent.end(), 1440, false);
+            if (newSlot.duration() > 0 && newSlot.duration() >= request.getDuration()) {
+                availableTimes.add(newSlot);
+            }
+        } 
+   
+    }
+
+    if (timeLeft.duration() > 0 && timeLeft.duration() >= request.getDuration()) {
+        availableTimes.add(timeLeft);
+    }
+
+    return availableTimes;
+  }
+
+  boolean checkDuplicateAttendees(Collection<String> s1, Set<String> s2) {
+      Iterator<String> iterator = s1.iterator();
+ 
+        while (iterator.hasNext()) {
+            String currPerson = iterator.next();
```

Comment:
Using the standard Java foreach loop of `for (String person : s1) {...}` is preferred whenever possible.

Reply:
ok!